### ISSUE TITLE
StreamIds are allocated in the order in which they are first polled, rather than opened

### ISSUE THREAD

**Issue by alecmocatta** [djc/quinn#660]
The docs specify incoming streams as being "in the order they were opened". Intuitively I'd expect this to be the order in which `open_uni`/`open_bi` were called, but it's actually the order in which they were first polled.

Ordering according to `open_*` is more explicit, the programmer has greater control over it (e.g. [`join_all`](https://docs.rs/futures/0.3/futures/future/fn.join_all.html) doesn't specify in which order it polls futures), plus I believe it to be less surprising. Is this worth resolving?

**Comment by Ralith** [comment]
The general rule in rust is that futures do nothing until they're polled. I agree that the resulting behavior could catch someone by surprise, though I expect the overwhelming majority of cases where someone cares about stream ordering will involve serial `await`ing.

A fix could be implemented by modifying `quinn_proto::Connection::open` to always successfully return a new ID, and introduce a separate mechanism to poll whether a stream is actually open yet. The open futures would then allocate an ID synchronously, and wait for it to become open before yielding the resulting handle.

**Comment by djc** [comment]
Would that allow opening streams (as the peer sees it) in a different order than what IDs are allocated? I worry that that might also be confusing in different ways.

**Comment by Ralith** [comment]
Remotely-initiated streams are always observed opening strictly in ID order. The "poll whether a stream is open" mechanism would be nothing more than `stream_id.index() < self.streams.max[stream_id.dir()]`.

**Comment by djc** [comment]
> Remotely-initiated streams are always observed opening strictly in ID order.

Is that true for other implementations?

**Comment by Ralith** [comment]
Yes, it's a property of the spec. Even if you receive stream frames out of order, receiving a higher-numbered stream always implicitly opens the lower-numbered ones first:

> [Before a stream is created, all streams of the same type with lower-numbered stream IDs MUST be created. This ensures that the creation order for streams is consistent on both endpoints](https://quicwg.org/base-drafts/draft-ietf-quic-transport.html#section-3.2-6).

**Comment by lmpn** [comment]
This issue is old is it still valid?

**Comment by Ralith** [comment]
Yes, this issue is current. The problem it describes is that outgoing stream IDs are assigned based on the order in which the stream opening futures are first polled, which could be surprising in certain rare cases. Typically people `await` the open futures serially, so it doesn't come up.
